src/base/controller.hpp

```diff
@@ -168,18 +167,22 @@ class Controller {
   //-----------------------------------------------------------------------
 
   // Set OpenMP thread settings to default values
-  void set_threads_default();
+  void clear_parallelization();
 
-  // Internal counter of number of threads still available for subthreads
-  int available_threads_ = 1;
+  // Set OpenMP thread settings for experiments
+  virtual void set_parallelization(Qobj& qobj);
 
   // The maximum number of threads to use for various levels of parallelization
-  // set to 0 for maximum available
-  int max_threads_total_;
-  int max_threads_circuit_;
-  int max_threads_shot_;
-  int max_threads_state_;
+  int max_parallel_threads_;
 
+  // Parameters for parallelization management in configuration
+  int max_parallel_experiments_;
+  int max_parallel_shots_;
+
+  // Parameters for parallelization management for experiments
+  int parallel_experiments_;
+  int parallel_shots_;
+  int parallel_gates_;
```

Comment:
I think we should change the variable name `parallel_gates_` as it makes it sound like the gates are done in parallel which isn't the case. We could use something else like `parallel_linalg_`, `parallel_state_update_` instead.